### DISCUSSION TITLE
Can I set non-changing build id in next.config.js?

### DISCUSSION THREAD

**Discussion by moaazbhnas228** [vercel/next.js#53346] [Help]
### Summary

We have an issue (mostly with Safari browsers) that the browser fetches data from an old build id which results in `Failed to load static props` error.
For example, here it fetches on the build id: `jXsY1IoTQdFEej40D_tdc`
whereas the build id we currently have in production is: `No2dz7NzLxHiQFHqc4oDQ`
![image](https://github.com/vercel/next.js/assets/111786023/93b9f3be-bf6f-4c55-80e1-bef0e14487a4)

I tried everything to solve this issue, but failed.
So I thought if I had a non-changing build id, this issue wouldn't occur again.

Does that cause any other issue that I should consider?

### Additional information

```
    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.16.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.9
      eslint-config-next: 13.4.3
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.8.2
    Next.js Config:
      output: N/A
```

**Comment by icyJoseph** [comment]
To control the build-id: https://nextjs.org/docs/app/api-reference/next-config-js/generateBuildId

```ts
module.exports = {
  generateBuildId: async () => {
    // You can, for example, get the latest git commit hash here
    return 'my-build-id'
  },
}
```

I think there's some implications to setting as a constant, but I can't find much recent docs about it:

> Next.js uses a constant id generated at build time to identify which version of your application is being served. This can cause problems in multi-server deployments when next build is run on every server. In order to keep a consistent build id between builds you can provide your own build id.

**Reply by moaazbhnas228** [reply]
Thank you, @icyJoseph 
I actually came from [generateBuildId](https://nextjs.org/docs/app/api-reference/next-config-js/generateBuildId) doc.
As you said there's no much information there.
I set it to a constant anyway, and let's see what happens 😅

**Reply by moaazbhnas228** [reply]
Tried it. Terrible idea

**Reply by icyJoseph** [reply]
Uff, what happened? Might be worth to document the consequences/implications.

**Reply by piersolenski** [reply]
@moaazbhnas228 keen to know as well, having the same issue!

**Reply by moaazbhnas228** [reply]
the website opens on the white "500" page. Don't know why ..

**Comment by y-a-v-a** [comment]
Having a non-changing build ID is a wrong approach, I think. As you might run into weirder issues because Safari might have an HTML page rendered by build `X` and receive a data file from build `Y`, not knowing it is from a new build. Most probably that causes `500` pages.
According to [this discussion](https://github.com/vercel/next.js/discussions/14726) it is intended behaviour that NextJS returns a 404, as the client side router seems to do a hard refresh of the page, making the client receive a complete page from build `Y`.

The only thing I do not like is the amount of 404s in my server logging, which I need to neglect but clutters other important messages.
Just an idea for a solution: a listener on `pageshow` could do a kind of pre-flight request for the latest `BUILD_ID` and compare it with the one stored on the client side, and if they're not identical, a hard refresh could be made...?

**Reply by moaazbhnas228** [reply]
Makes sense. Will try and let you know.
Thank you

**Reply by silviow** [reply]
@moaazbhnas228 Hey! Have you made progress on this topic? 👀 

Could you share if you found a solution, please? :)

**Reply by moaazbhnas228** [reply]
Hi @silviow,
What reduced the error for me was this advice: [/next.js/issues/43088#issuecomment-1408695026](https://github.com/vercel/next.js/issues/43088#issuecomment-1408695026).
I created a handler error page "[...myErrorPage]" where all unmatched routes land on.
In this page, I perform a hard refresh every 4 seconds. This should fetch the new version.
Also in Next.js 404 page, all links crash for some reason, but not in [...myErrorPage].
I got a lot less complains from customers after going with this approach.

``` js
const router = useRouter();

  const handleRefresh = () => {
    router.reload();
  };

  useEffect(() => {
    const timeoutId = setTimeout(handleRefresh, 4000);
    return () => clearTimeout(timeoutId);
  }, []);
```
